tests/State/ScopeTest.php

```diff
@@ -0,0 +1,200 @@
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sentry\Tests\State;
+
+use PHPUnit\Framework\TestCase;
+use Sentry\Breadcrumbs\Breadcrumb;
+use Sentry\Configuration;
+use Sentry\Event;
+use Sentry\Severity;
+use Sentry\State\Scope;
+
+final class ScopeTest extends TestCase
+{
+    public function testSetTag(): void
+    {
+        $scope = new Scope();
+
+        $this->assertEquals([], $scope->getTags());
+
+        $scope->setTag('foo', 'bar');
+        $scope->setTag('bar', 'baz');
+
+        $this->assertEquals(['foo' => 'bar', 'bar' => 'baz'], $scope->getTags());
+    }
+
+    public function testSetExtra(): void
+    {
+        $scope = new Scope();
+
+        $this->assertEquals([], $scope->getExtra());
+
+        $scope->setExtra('foo', 'bar');
+        $scope->setExtra('bar', 'baz');
+
+        $this->assertEquals(['foo' => 'bar', 'bar' => 'baz'], $scope->getExtra());
+    }
+
+    public function testSetUser(): void
+    {
+        $scope = new Scope();
+
+        $this->assertEquals([], $scope->getUser());
+
+        $scope->setUser(['foo' => 'bar']);
+
+        $this->assertEquals(['foo' => 'bar'], $scope->getUser());
+
+        $scope->setUser(['bar' => 'baz']);
+
+        $this->assertEquals(['bar' => 'baz'], $scope->getUser());
+    }
+
+    public function testSetFingerprint(): void
+    {
+        $scope = new Scope();
+
+        $this->assertEmpty($scope->getFingerprint());
+
+        $scope->setFingerprint(['foo', 'bar']);
+
+        $this->assertEquals(['foo', 'bar'], $scope->getFingerprint());
+    }
+
+    public function testSetLevel(): void
+    {
+        $scope = new Scope();
+
+        $this->assertNull($scope->getLevel());
+
+        $scope->setLevel(Severity::debug());
+
+        $this->assertEquals(Severity::debug(), $scope->getLevel());
+    }
+
+    public function testAddBreadcrumb(): void
+    {
+        $scope = new Scope();
+        $breadcrumb1 = new Breadcrumb(Severity::error(), Breadcrumb::TYPE_ERROR, 'error_reporting');
+        $breadcrumb2 = new Breadcrumb(Severity::error(), Breadcrumb::TYPE_ERROR, 'error_reporting');
+        $breadcrumb3 = new Breadcrumb(Severity::error(), Breadcrumb::TYPE_ERROR, 'error_reporting');
+
+        $this->assertEmpty($scope->getBreadcrumbs());
+
+        $scope->addBreadcrumb($breadcrumb1);
+        $scope->addBreadcrumb($breadcrumb2);
+
+        $this->assertSame([$breadcrumb1, $breadcrumb2], $scope->getBreadcrumbs());
+
+        $scope->addBreadcrumb($breadcrumb3, 2);
+
+        $this->assertSame([$breadcrumb2, $breadcrumb3], $scope->getBreadcrumbs());
+    }
+
+    public function testAddEventProcessor(): void
+    {
+        $callback1Called = false;
+        $callback2Called = false;
+        $callback3Called = false;
+
+        $event = new Event(new Configuration());
+        $scope = new Scope();
+
+        $scope->addEventProcessor(function (Event $eventArg) use (&$callback1Called): ?Event {
```

Comment:
Should we test in some way the order of invocation of the callbacks?

Reply:
Seems fair, any idea how to do it?

Reply:
You probably have to pass all the booleans to all the callbacks and check true/false accordingly.